## Problem 1. 

(a)
$A = \begin{pmatrix}
1 & 0 & 0 \\
-1 & 0 & 0 \\
0 & 1 & 0 \\
0 & -1 & 0 \\
0 & 0 & 1 \\
0 & 0 & -1 
\end{pmatrix}$
  $b^T = (1, 1,1, 1,1, 1,1, 1)$ 


(b)
$A = \begin{pmatrix}
1 & 1 & 1 \\
-1 & -1 & -1 \\
-1 & 1 & 1 \\
1 & -1 & -1 \\
1 & -1 & 1 \\
-1 & 1 & -1 \\
1 & 1 & -1 \\
-1 & -1 & 1 
\end{pmatrix}$
  $b^T = (1, 1,1, 1,1, 1,1, 1, 1, 1)$

## Problem 2.

## (a)

### Variable

Transfer the bounded variables to nonnegative variables:

Let $z_1$ to be $z_1 = x_1 - x_2$, where $x_1$ and $x_2$ are nonnegative variables.

Let $z_2$ to be $z_2 = x_3 -1$, then $0 \leq x_3 \leq 6$

Let $z_3$ to be $z_3 = x_4 -1$, then $0 \leq x_4 \leq 6$

Let $z_4$ to be $z_4 = x_5 -2$, then $0 \leq x_5 \leq 4$

Let $x_6$ to be the surplus variable 

Let $x_7, x_8, x_9, x_{10}$ to be the slack variable

### Objective

Transfer maximize to minimize:

$Minimize -p = -(3z_1 - z_2)$, Which is equal to 

$Minimize -p = -3x_1 + 3x_2 + x_3 -1$

In standard form, we write $Minimize \, -p+1 = -3x_1 + 3x_2 + x_3$, where the objective value of orginal problem is $p$
### Constraints

The constraints can be written as:

$
\begin{align*}
Minimize& -3x_1 + 3x_2 + x_3  \\
Subject \, to& -x_1 + x_2 + 6x_3 -x_4 +x_5 - x_6 &= &4  \\
&7x_3 +x_5 &= &14  \\
&x_4 + x_5 + x_7 &=&5  \\
&x_3 + x_8 &=&6  \\
&x_4 + x_9 &=&6  \\
&x_5 + x_{10} &=&4  \\
&&x_1, ... ,x_{10} \geq 0 \\
\end{align*}
$

$A = \begin{pmatrix}
-1 & 1 & 6 & -1 & 1 & -1 & 0 & 0 & 0 & 0 \\
0 & 0 & 7 & 0 & 1 & 0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 1 & 1 & 0 & 1 & 0 & 0 & 0 \\
0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 & 0 \\
0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1 & 0 \\
0 & 0 & 0 & 0 & 1 & 0 & 0 & 0 & 0 & 1  
\end{pmatrix}$

$b^T = (4,14,5,6,6,4)$

$c^T = (-3,3,1,0,0,0,0,0,0,0)$

$x^T = (x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_{10})$

## (b)

In [59]:
using JuMP, Clp
lb=[-Inf,-1,-1,-1]
ub=[Inf,5,5,2]
m = Model(solver=ClpSolver())
@variable(m, z[i=1:4], lowerbound=lb[i], upperbound=ub[i])
@constraint(m, -z[1] + 6z[2] - z[3] + z[4] >= -3 )
@constraint(m, 7z[2] + z[4] == 5 )
@constraint(m, z[3] + z[4] <= 2 )
@objective(m, Max, 3z[1] - z[2] )

status = solve(m)

println(m)
println(status)
println("z = ", getvalue(z))
println("objective = ", getobjectivevalue(m) )


Max 3 z[1] - z[2]
Subject to
 -z[1] + 6 z[2] - z[3] + z[4] ≥ -3
 7 z[2] + z[4] = 5
 z[3] + z[4] ≤ 2
 z[1]
 -1 ≤ z[2] ≤ 5
 -1 ≤ z[3] ≤ 5
 -1 ≤ z[4] ≤ 2

Optimal
z = [8.57143, 0.428571, -1.0, 2.0]
objective = 25.28571428571429


In [194]:
A = [
    -1 1 6 -1 1 -1 0 0 0 0; 
    0 0 7 0 1 0 0 0 0 0 ; 
    0 0 0 1 1 0 1 0 0 0 ; 
    0 0 1 0 0 0 0 1 0 0 ; 
    0 0 0 1 0 0 0 0 1 0 ; 
    0 0 0 0 1 0 0 0 0 1 
]
b = [4; 14; 5; 6; 6; 4]
c = [-3;3;1;0;0;0;0;0;0;0]
using JuMP, Clp

m = Model(solver=ClpSolver())
@variable(m, x[1:10] >= 0 )     
@constraint(m, A*x .== b )  
@objective(m, Min, dot(c,x) )   
status = solve(m)

println(m)
println(status)
println()
println("z[1] = ", getvalue(x[1])-getvalue(x[2]) )
println("z[2] = ", getvalue(x[3])-1 )
println("z[3] = ", getvalue(x[4])-1 )
println("z[4] = ", getvalue(x[5])-2 )
println("Objective= ", -(getobjectivevalue(m)-1))



Min -3 x[1] + 3 x[2] + x[3]
Subject to
 -x[1] + x[2] + 6 x[3] - x[4] + x[5] - x[6] = 4
 7 x[3] + x[5] = 14
 x[4] + x[5] + x[7] = 5
 x[3] + x[8] = 6
 x[4] + x[9] = 6
 x[5] + x[10] = 4
 x[i] ≥ 0 ∀ i ∈ {1,2,…,9,10}

Optimal

z[1] = 8.571428571428571
z[2] = 0.4285714285714286
z[3] = -1.0
z[4] = 2.0
Objective= 25.28571428571429


## Problem 3.

Let $\textbf{x}$ denote the amount of each raw material used to produce the steel. 



In [70]:
using JuMP, Clp
lb=[0,0,0,0,0,0,0]
ub=[400,300,600,500,200,300,250]
c=[200,250,150,220,240,200,165]
m = Model(solver=ClpSolver())
@variable(m, x[i=1:7], lowerbound=lb[i], upperbound=ub[i])
@constraint(m, sum(x) == 500)
@constraint(m, 2*500 <= 2.5x[1] + 3x[2] <= 3*500 )
@constraint(m, 0.4*500 <= 0.3x[3] + 90x[4] + 96x[5] + 0.4x[6] + 0.6x[7]<= 0.6*500 )
@constraint(m, 1.2*500 <= 1.3x[1] + 0.8x[2] + 4x[5] + 1.2x[6]<= 1.65*500 )
@objective(m, Min, c'*x )

status = solve(m)

println(m)
println(status)
println("x = ", getvalue(x))
println("objective = ", getobjectivevalue(m) )


Min 200 x[1] + 250 x[2] + 150 x[3] + 220 x[4] + 240 x[5] + 200 x[6] + 165 x[7]
Subject to
 x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7] = 500
 1000 ≤ 2.5 x[1] + 3 x[2] ≤ 1500
 200 ≤ 0.3 x[3] + 90 x[4] + 96 x[5] + 0.4 x[6] + 0.6 x[7] ≤ 300
 600 ≤ 1.3 x[1] + 0.8 x[2] + 4 x[5] + 1.2 x[6] ≤ 825
 0 ≤ x[1] ≤ 400
 0 ≤ x[2] ≤ 300
 0 ≤ x[3] ≤ 600
 0 ≤ x[4] ≤ 500
 0 ≤ x[5] ≤ 200
 0 ≤ x[6] ≤ 300
 0 ≤ x[7] ≤ 250

Optimal
x = [400.0, 0.0, 39.7763, 0.0, 2.76127, 57.4624, 0.0]
objective = 98121.63579168123


The composition is 400.0, 0.0, 39.7763, 0.0, 2.76127, 57.4624, 0.0 tons for each raw material

## Problem 4 

## (a)
Let x denote the amount of food (daily)

In [1]:
# STARTER CODE FOR STIGLER'S DIET PROBLEM
using NamedArrays

# import Stigler's data set
raw = readcsv("stigler.csv")
(m,n) = size(raw)

n_nutrients = 2:n      # columns containing nutrients
n_foods = 3:m          # rows containing food names

nutrients = raw[1,n_nutrients][:]   # the list of nutrients (convert to 1-D array)
foods = raw[n_foods,1][:]           # the list of foods (convert to 1-D array)

# lower[i] is the minimum daily requirement of nutrient i.
lower = Dict( zip(nutrients,raw[2,n_nutrients]) )

# data[f,i] is the amount of nutrient i contained in food f.
data = NamedArray( raw[n_foods,n_nutrients], (foods,nutrients), ("foods","nutrients") );

In [2]:
using JuMP, Clp

m = Model(solver=ClpSolver())
@variable(m, x[1:size(data)[1]] >= 0 ) 
@constraint(m, convert(Array{Float64,2},data)'*x .>= raw[2,n_nutrients])  
@objective(m, Min, 365*sum(x) )   
status = solve(m)

println(status)
println()
println("x = ", getvalue(x) )
println()
println("Objective= ", getobjectivevalue(m))
println()
println("The foods make up the diet is")
for i = 1:77 
    if getvalue(x[i])>0 
        println("No.",i," ",foods[i], " ", getvalue(x[i]))
    end 
end

Optimal

x = [0.0295191, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00189256, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0112144, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00500766, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0610286, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

Objective= 39.66173154546625

The foods make up the diet is
No.1 Wheat Flour (Enriched) 0.02951906167648827
No.30 Liver (Beef) 0.0018925572907052643
No.46 Cabbage 0.011214435246144865
No.52 Spinach 0.005007660466725203
No.69 Navy Beans, Dried 0.061028563526693246


## (b)

In [4]:
n=m
@constraint(n, x[30]==0)
status = solve(n)

println(status)
println()
println("x = ", getvalue(x) )
println()
println("Objective= ", getobjectivevalue(n))
println()
println("The foods make up the diet is")
for i = 1:77 
    if getvalue(x[i])>0 
        println(foods[i], " ", getvalue(x[i]))
    end 
end

Optimal

x = [0.0354556, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00859146, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0112495, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00511283, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.048628, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

Objective= 39.79866435040896

The foods make up the diet is
Wheat Flour (Enriched) 0.03545558140888772
Evaporated Milk (can) 0.008591461668763572
Cabbage 0.011249517312443502
Spinach 0.005112832613199645
Navy Beans, Dried 0.048628043573168474
